src/backend/distributed/planner/distributed_planner.c

```diff
@@ -794,6 +811,171 @@ CreateDistributedPlan(uint64 planId, Query *originalQuery, Query *query, ParamLi
 }
 
 
+static void
+InlineCTEs(Query *query)
+{
+	ListCell *cteCell = NULL;
+	CmdType cmdType = query->commandType;
+
+	List *copiedList = list_copy(query->cteList);
+
+	foreach(cteCell, copiedList)
+	{
+		CommonTableExpr *cte = (CommonTableExpr *) lfirst(cteCell);
+
+		if ((cte->ctematerialized == CTEMaterializeNever ||
```

Comment:
I think at this point we can do `SubqueryColocated(query, cte->ctequery)` to prevent unncessary cte inlining. Because if the query wouldn't yield a colocated one, there is no need to inline the CTE.

That could even convert some intermediate result joins to repartition joins, which we probably wouldn't want at this point.